### PR TITLE
Fixes #31132: Downcase MAC addresses

### PR DESCRIPTION
**Pull request by neomilium**
In order to prevent from the side effects of changing MAC address from upcased to downcase (as explained in Redmine issue), this patch always return a downcased MAC address.

### CODE REVIEW

File: app/models/nic/interface.rb
Comment (lzap) [inline app/models/nic/interface.rb]: Please create two methods named `mac_downcase` and `mac_upcase` and then add all two methods into safemode whitelist. Then use whatever you need in your template. I would like to avoid change of behavior.

**Review by lzap — CHANGES_REQUESTED** [review]
Thanks for the patch, I'd like to add two explicit helpers for that rather than changing the current behavior.

File: app/models/nic/interface.rb
Comment (neomilium) [reply app/models/nic/interface.rb]: The behavior change is expected. Creating `mac_downcase` and `mac_upcase` is useless as we can already use `mac.downcase` and `mac.upcase` in templates.

It seems my issue explanation was not enough clear: my troubles comes from the MAC address change during machine life-cycle due to switch between upcased and downcased.

From iPXE to finish template, the process is disturbed by MAC changes.

Comment (neomilium) [reply app/models/nic/interface.rb]: Let me know if clearer scenario is require to understand the issues.

Comment (lzap) [reply app/models/nic/interface.rb]: Oh I understand now. Please implement an ActiveRecord before save hook which converts mac to downcase prior saving the record. You can also implement a record validator which refuses to store record which has uppercase mac address. If you need to fix all existing records, create new Rails migration which makes sure all mac addresses are in lowercase.

Comment (lzap) [reply app/models/nic/interface.rb]: Oh we already have that in our codebase: `normalize_mac` we just call it to do validation. Implement `mac=` method similarly we do for `ip=` etc. Add a test for this.

Comment (lzap) [reply app/models/nic/interface.rb]: Hello, I hope you're doing well. My last comment is blocking this from to be merged.

Comment (neomilium) [reply app/models/nic/interface.rb]: Hello @lzap your advises are perfect. I'm working on another topic and will be back on this one in few weeks.

I'll turn this PR in draft.

Thanks!

**Comment by mtancoigne** [comment]
After a first step into this, I wonder if we should only downcase the MAC in the `mac=` method, or if we also should normalize it:

- Downcase only don't prevent for invalid addresses until `normalize_mac` is called (before validation)
- Normalize in `mac=` leads to the question of "how do we handle invalid values?". Possibilities include
  - raise an error (and possibly break mass assignments)
  - fail silently and set the value to nil or the old value, without notification (which is problematic)
  - ??

I can't figure a "right way" to do this without changing how Foreman behaves: we use the model before it's validated, so we may use it with incorrect values, and I guess we can't validate unless we performed some actions on it...
Maybe the less intrusive way to normalize mac addresses case for now is to only downcase in `mac=`, leaving the possibility to _use_ invalid mac adresses before the model is validated and saved.

What do you think ?

**Comment by neomilium** [comment]
Even `downcase` reach our goal, `normalize_mac` makes more sense.

Raising an error seems to me a better idea but I can't figure out the case where we need to do mass assignments.

@lzap Do you have a usecase where raising an error will be a problem?

**Comment by lzap** [comment]
Well my general expectation would be to perform normalization during before_save but if there's something that needs this earlier (we do a LOT of stuff during validation) then do it via validator, that's okay.

About the question of what to do, general expectation would be: if quack like MAC, convert it to valid MAC (e.g. downcase, replace `:` with `-` or whatever we do), otherwise fail with an error.

**Comment by mtancoigne** [comment]
I'll work on this in the next days.

**Comment by mtancoigne** [comment]
The fix proposes to normalize the mac when it's defined, and it will throw an error when an invalid mac is used. I'm in the process of updating the tests.

File: db/migrate/20210407104300_downcase_mac_addresses.rb
Comment (ekohl) [inline db/migrate/20210407104300_downcase_mac_addresses.rb]: Isn't there a pure SQL way to do this? Untested, but perhaps this works?
```suggestion
    Nic::Managed.where.not(mac: nil).update_all('mac = lower(mac)')
```

Comment (mtancoigne) [reply db/migrate/20210407104300_downcase_mac_addresses.rb]: Even better: `execute 'UPDATE nics SET mac=LOWER(mac)'`

**Comment by mtancoigne** [comment]
I _think_ all the tests involving mac addresses were updated; I got a lot of failures on unrelated stuff (mainly integration errors).

File: test/controllers/api/v2/hosts_controller_test.rb
Comment (ekohl) [inline test/controllers/api/v2/hosts_controller_test.rb]: This looks like a bad change. Invalid data should not result in an internal server error. That's why it was using a validation.

File: db/migrate/20210407104300_downcase_mac_addresses.rb
Comment (ekohl) [inline db/migrate/20210407104300_downcase_mac_addresses.rb:3]: I'd like to as a more experienced Rails dev to comment on the right way to do this. I'd say that using the ORM is normally a good thing but maybe migrations are different.

Comment (mtancoigne) [reply db/migrate/20210407104300_downcase_mac_addresses.rb:3]: Using the ORM in migration can be dangerous: if for any reason the model vanishes, the migration is broken.

File: test/controllers/api/v2/hosts_controller_test.rb
Comment (mtancoigne) [reply test/controllers/api/v2/hosts_controller_test.rb]: Sorry for the long time, I had not seen this comment. 

I agree this change don't stick with a "normal" API behavior. As Foreman uses host values _before_ validation we need to find a way to validate without throwing an exception, but silently ignoring values is a no-go too. The simplest solution is either to downcase the mac when the attribute is _set_ or when it's _get_, and leave the possibility to set incorrect values until it's validated (e.g. : in a11dd500d1f164744db1f06ab38e8b86eabe97a7)

Better solutions are welcome!

Comment (neomilium) [reply test/controllers/api/v2/hosts_controller_test.rb]: @mtancoigne How to know what operations are made by Foreman _before_ validation?

Comment (mtancoigne) [reply test/controllers/api/v2/hosts_controller_test.rb]: It's hard to tell precisely without digging deep; at first sight, [the compute profile is applied](https://github.com/theforeman/foreman/blob/develop/app/controllers/api/v2/hosts_controller.rb#L141). The issue we try to address originally is the use of mixed upcased/downcased mac adresses somewhere in the steps of a virtual machine creation.

The mac addresses [being downcased](https://github.com/theforeman/foreman/blob/develop/lib/net/validations.rb#L122) [_before_ validation](https://github.com/theforeman/foreman/blob/develop/app/models/nic/base.rb#L14) lets me suppose that host values are used between instanciation and validation.

File: db/migrate/20210407104300_downcase_mac_addresses.rb
Comment (lzap) [reply db/migrate/20210407104300_downcase_mac_addresses.rb:3]: I am far from an experienced RoR dev, but generally I like this because we only support PostgreSQL and this will be much faster than via ORM. Our biggest customers have up to 100,000 nodes and that is not many records to update, but some customers do have many NICs in their database created by accident (unfiltered container hosts) and then amount of records could be actually higher.

File: app/models/nic/base.rb
Comment (lzap) [inline app/models/nic/base.rb]: What exactly wrong is with before validation? We use Rails callbacks a lot through the codebase, teach me.

File: test/controllers/api/v2/hosts_controller_test.rb
Comment (lzap) [reply test/controllers/api/v2/hosts_controller_test.rb]: I think you can split validation of mac and converting it to downcase into two validation methods. One that validates should raise error while the downcase should only downcase (if its a string).

**Comment by lzap** [comment]
I am okay with the migration, but I would stick with validation callback rather than the "super" hack.

File: test/controllers/api/v2/hosts_controller_test.rb
Comment (mtancoigne) [reply test/controllers/api/v2/hosts_controller_test.rb]: If there are validators, they won't be triggered before the validation step and the current behavior won't change. If the downcase should only downcase, the best place for it is still in the `mac=` method, so it's downcased on assignation, right ?

File: app/models/nic/base.rb
Comment (mtancoigne) [reply app/models/nic/base.rb]: Rails model hooks cycle is as follow:

1. instancation (`new MyModel param: value, param2: value2, ...`)
2. _before_validation_: model attributes may be wrong and can be normalized here
3. validation
4. _before_save_ (maybe "after validation too"): model passed validation and is ready to be saved
5. save
6. _after_save_: save was successfull.

As we can see, using model data while validation did not happen is a risk to use invalid values in another part of the application (simple example of this:

```rb
post = Post.new title: '       some badly formatted garbage'
post.send_notification_emails
post.save # will trigger "before_validation" hooks which in this example cleans the title, but mails are sent using the wrong one.
```

The current Foreman's behavior about mac addresses is to normalize it on a _before_validation_ hook, which is fine, as long as the data is not used somewhere else, but as we encountered issues with mac addresses being upcased in the steps of a virtual machine creation, our guess was an issue with the mac being used before its validation (and normalization).

The minimal fix to this is to downcase macs in the database for consistency and downcase macs when they are assigned to the object (the `mac=` method).

The best approach would be to manage to perform validation earlier in the process (fail fast!) but this will need a good understanding of the mechanisms involved behind host creation, in order to reorder the sequence. It would be something like:

```rb
host Host.new(params)
host.validate
if host.valid?
  continue_machine_creation_process
else
  # raise or return the errors to the user
end
```

If things are done in the model, it's a better idea to do it in the `before_save` hook, as data should be valid at this point.

But as I say, this rework would need a good understanding and some time...

I hope these explanations helped!

Comment (neomilium) [reply app/models/nic/base.rb]: @mtancoigne Many thanks for detailed explanations.
@lzap @ekohl After this explanation, what do you suggest?
IMHO, 
 * API break is a no-go,
 * reorganize whole validation/normalization process requires a mentor and probably a sponsorship as its time consuming if we want to refactor this safely
 * the left way is to use the first tiny/poor patch I provide at first with the db migration added

Comment (mtancoigne) [reply app/models/nic/base.rb]: I agree with that, if we don't want to rework too much code, the small patch seems to be the solution (overriding `mac=` to downcase the value).

Comment (ekohl) [reply app/models/nic/base.rb]: It may be ok to override `mac=` but `normalize_mac` can raise an error if it's invalid. That's probably the main reason that the API change is there. Perhaps there needs to be a best effort normalize in the assignment while still allowing invalid data. Then a validation that does proper error handling.

And yes, the whole host orchestration is one area that deserves a lot of love. it can be very complex because a compute resource may provide a MAC address. That may be what you're running into.

I think that's happening here:
https://github.com/theforeman/foreman/blob/ceb276fbe96b97770f5d292b1eadca0205a34a0a/app/models/concerns/orchestration/compute.rb#L347-L349

That may also be a place to normalize, but I think it's wrong because it may lead to a mismatch. It's probably better for a compute provider to solve it so it always returns a good value. For example here: https://github.com/theforeman/foreman_fog_proxmox/blob/ca9475c15731e866cc3d8b1998320b30838f16d1/app/models/concerns/fog_extensions/proxmox/server.rb#L47-L49

What do you think?

Comment (lzap) [reply app/models/nic/base.rb]: My guts are telling me this is not right, but I am not that skilled RoR developer so I am gonna pull @ares or @ezr-ondrej to assist here. What I particularly don't like is that this is not a problem of MAC field, but generic flaw of Ruby on Rails design. Also where do get the confidence that somewhere deep in Rails code it will use attribute setter to set the MAC instead of accessing an instance variable from?

I feel like fixing the MAC address in both UI and API controller is way cleaner - these are the main (perhaps all) input points where a human would enter a MAC address. But I am not the one to decide here. I do appreciate the effort and I do understand the importance of this bug to be fixed, thank you for patience with us.

File: db/migrate/20210407104300_downcase_mac_addresses.rb
Comment (ezr-ondrej) [reply db/migrate/20210407104300_downcase_mac_addresses.rb:3]: Usually using the ORM class is better, then direct SQL query, because individual setups can choose to use prefixed table names, but that is not an issue for us, so this is ok to use.

the best for this is usually:
```
Nic::Base.update_all(mac: Nic::Base.arel_table[:mac].lower)
```
There is quite some magic, that rails does and can be useful, but none of that is used here I'd bet.

The case where class dissapear is valid, but in that case you want the migration to fail, so you know what migrations were touching that class table so you can deal with them :))

File: app/models/nic/base.rb
Comment (ezr-ondrej) [reply app/models/nic/base.rb]: I believe this is flaw of our orchestration model, that is doing FAAAR too much in `around_save` and most of it `before_save` hook, where as @ekohl pointed out we allow changing the model (that's antipatern because it is after the validation have passed)
Two issues I see with this:
1. We can fail with exception on setting invalid value, that is quite bad
2. We do not validate the mac address anymore.

I agree with @ekohl probably the easiest would be normalizing in the setter the best effort (probably just the `.downcase` would be enough). Plus keep the validation on mac address.

File: db/migrate/20210407104300_downcase_mac_addresses.rb
Comment (mtancoigne) [reply db/migrate/20210407104300_downcase_mac_addresses.rb:3]: I reverted; I still need to check for all the tests.

**Review by lzap — DISMISSED** [review]
I am good, final call is on others here, I am a bit fuzzy in Rails here.

**Comment by ezr-ondrej** [comment]
ok to test

**Comment by ezr-ondrej** [comment]
Solving this per provider would be better, but I think it should be fine like this, I think dhcp is not case sensitive on mac, but not sure there.
If comparison of mac is case sensitive somewhere it is bug and should be solved there.

Thus I'm ok with this change, @neomilium can you undraft this please?

File: test/models/host_test.rb
Comment (ekohl) [inline test/models/host_test.rb]: Technically you could even leave this uppercase and verify that it's downcased, right?

File: test/controllers/api/v2/hosts_controller_test.rb
Comment (neomilium) [reply test/controllers/api/v2/hosts_controller_test.rb]: Now outdated, closing conversation.

File: app/models/nic/base.rb
Comment (neomilium) [reply app/models/nic/base.rb]: OK for this, it resolves conversation.

File: test/models/host_test.rb
Comment (neomilium) [reply test/models/host_test.rb]: Thanks, I revert this line.

**Comment by ezr-ondrej** [comment]
[test katello]

**Review by ezr-ondrej — APPROVED** [review]
LGTM :+1:

**Comment by ezr-ondrej** [comment]
Thanks @neomilium !
I believe we should address the core issue that we are touching the mac after the validation, but it is out of scope as that would mean rewrite orchestration.

**Pull request merged**
